### PR TITLE
1.0 controller

### PR DESCRIPTION
**Pull request by emulgeator**

### CODE REVIEW

File: src/Controller/ControllerAbstract.php [hunk 1]
```diff
@@ -36,162 +22,127 @@ abstract class ControllerAbstract implements IController
     /** @var IResponse */
     protected $response;
```
Comment (szeber) [inline src/Controller/ControllerAbstract.php]: I'm thinking this and the request should be private and have a protected getter instead. The reason is, this is a constant source of pain when using traits.

Comment (szeber) [reply src/Controller/ControllerAbstract.php]: Or it can stay protected, but have a getter...

File: src/Controller/ControllerAbstract.php [hunk 2]
```diff
@@ -36,162 +22,127 @@ abstract class ControllerAbstract implements IController
...
 
-    public function __construct()
-    {
-        Application::getInstance()->getDiContainer()->getViewDo()->clear();
-    }
-
-    public function setRequest(IRequest $request): void
+    public function setRequest(IRequest $request)
```
Comment (szeber) [inline src/Controller/ControllerAbstract.php]: A self typehint (with a $this docblock) would be great for these setters

File: src/Controller/HttpControllerAbstract.php [hunk 3]
```diff
@@ -14,102 +17,72 @@
 /**
  * Base class for HTTP controllers.
  */
-abstract class HttpController extends ControllerAbstract
+abstract class HttpControllerAbstract extends ControllerAbstract
```
Comment (szeber) [inline src/Controller/HttpControllerAbstract.php]: I'm not sure this class is required. The batches will look completely different. Basically all controllers will be http controllers.

File: src/Controller/HttpControllerAbstract.php [hunk 4]
```diff
@@ -14,102 +17,72 @@
...
-     * @param string $url        The URL to redirect to.
-     * @param int    $statusCode The status code of the redirect (3XX).
-     *
-     * @return void
-     *
-     * @throws \YapepBase\Exception\RedirectException   To stop execution of the controller.
+     * @throws RedirectException
      */
-    protected function redirectToUrl($url, $statusCode = 303)
+    protected function redirectToUrl(string $url, int $statusCode = 303): void
```
Comment (szeber) [inline src/Controller/HttpControllerAbstract.php]: Would be nice to use a constant for the status code. I think we already have a lib which lists all the status codes...

File: src/Controller/Rest/Exception/InvalidPayloadException.php [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+{
+    public function __construct(string $jsonDecodeError)
+    {
+        $message = 'Failed to decode the request as valid JSON. JSON decode error: ' . $jsonDecodeError;
+        parent::__construct($message);
+    }
+
+    public function getRecommendedHttpStatusCode(): int
+    {
+        return 400;
```
Comment (szeber) [inline src/Controller/Rest/Exception/InvalidPayloadException.php]: This should be 422

File: src/Controller/Rest/Exception/UnauthorizedException.php [hunk 6]
```diff
@@ -0,0 +1,22 @@
+<?php
+declare(strict_types=1);
+
+namespace YapepBase\Controller\Rest\Exception;
+
+class UnauthorizedException extends Exception
```
Comment (szeber) [inline src/Controller/Rest/Exception/UnauthorizedException.php]: We should have a 404 too if you have an endpoint that takes an ID and that ID is invalid too.

File: src/Controller/ControllerAbstract.php (hunk 2 above)
Comment (emulgeator) [reply src/Controller/ControllerAbstract.php]: See interface

File: src/Controller/HttpControllerAbstract.php (hunk 3 above)
Comment (emulgeator) [reply src/Controller/HttpControllerAbstract.php]: Ok I renamed the HttpRequest and HttpResponse as well.

File: src/Controller/HttpControllerAbstract.php (hunk 4 above)
Comment (emulgeator) [reply src/Controller/HttpControllerAbstract.php]: I tried it in another project but it's a pain in the @ss.
The problem is that as a developer you always think of the code first when you use it, and you always need to find the name of the constant. 
Even bigger problem is that the IDE wont help you at all, as it orders the codecompletion list by the constant name not by the value obviously.
So even if you are familiar with the status codes and you know exactly what to use when, you will need to learn the constant names.

File: src/Controller/Rest/Exception/InvalidPayloadException.php (hunk 5 above)
Comment (emulgeator) [reply src/Controller/Rest/Exception/InvalidPayloadException.php]: Just an example, this is called HTTP_UNPROCESSABLE_ENTITY.
The names are usually correct and meaningful, you will just always need to open the Httpstatuscodes class and find it first ;)

**Pull request merged**